zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java

```diff
@@ -1284,6 +1306,42 @@ private void updateParagraph(NotebookSocket conn, HashSet<String> userAndRoles,
     }
   }
 
+  private void patchParagraph(NotebookSocket conn, HashSet<String> userAndRoles,
+                              Notebook notebook, Message fromMessage) throws IOException {
+    String paragraphId = (String) fromMessage.get("id");
+    if (paragraphId == null) {
+      return;
+    }
+
+    String noteId = getOpenNoteId(conn);
+    if (noteId == null) {
+      noteId = (String) fromMessage.get("noteId");
+    }
+
+    if (!hasParagraphWriterPermission(conn, notebook, noteId,
+        userAndRoles, fromMessage.principal, "write")) {
+      return;
+    }
+
+    final Note note = notebook.getNote(noteId);
+    Paragraph p = note.getParagraph(paragraphId);
+
+    DiffMatchPatch dmp = new DiffMatchPatch();
+    String patchText = (String) fromMessage.get("patch");
+
+    // javascript add "," if change contains several patches
+    patchText = patchText.replace(",@@", "@@");
+    LinkedList<DiffMatchPatch.Patch> patches =
```

Comment:
`dm.patchFromText()` returns `List<DiffMatchPathc.Path>` type. Is there any specific reason to use `LinkedList`?

Reply:
patchApply(LinkedList<DiffMatchPatch.Patch> patches, String text)
First argunment requires LinkedList<DiffMatchPatch.Patch>